functionary/core/models/workflow_run.py

```diff
@@ -0,0 +1,145 @@
+        to="Workflow", on_delete=models.CASCADE, related_name="runs"
+    )
+    environment = models.ForeignKey(to="Environment", on_delete=models.CASCADE)
+    status = models.CharField(
+        max_length=16, choices=Task.STATUS_CHOICES, default=Task.PENDING
+    )
+    parameters = models.JSONField(blank=True, null=True, encoder=DjangoJSONEncoder)
+    creator = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.PROTECT)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    class Meta:
+        indexes = [
+            models.Index(
+                fields=["environment", "status"], name="wr_environment_status"
+            ),
+            models.Index(
+                fields=["environment", "creator"], name="wr_environment_creator"
+            ),
+            models.Index(
+                fields=["environment", "created_at"], name="wr_environment_created_at"
+            ),
+            models.Index(
+                fields=["environment", "updated_at"], name="wr_environment_updated_at"
+            ),
+        ]
+
+    @classmethod
+    def find_by_task(cls, task: Task) -> Optional["WorkflowRun"]:
+        """Retrieve the WorkflowRun that spawned a Task, if one exists
+
+        Args:
+            task: the Task for which to find the WorkflowRun
+
+        Returns:
+            WorkflowRun if the Task was spawned by one. Otherwise None."""
+        try:
+            return task.workflow_run_step.workflow_run
+        except ObjectDoesNotExist:
+            return None
+
+    def get_context(self) -> Context:
+        """Generates a context for resolving tasking parameters.
+
+        Returns:
+            A Context containing data from all WorkflowRunSteps that have
+            occurred for this WorkflowRun
+        """
+
+        # Malleable object class for storing context properties
+        class Object:
+            pass
+
+        context = {"parameters": Object()}
+
+        parameters = self.parameters or {}
+        for key, value in parameters.items():
+            setattr(context["parameters"], key, json.dumps(value))
+
+        for step in self.steps.all():
+            name = step.workflow_step.name
+            task = step.task
+
+            context[name] = Object()
+            context[name].result = task.raw_result
+
+        return Context(context)
+
+    def _update_status(self, status: str) -> None:
+        if self.status != status:
+            self.status = status
+            self.save()
+
+    def complete(self) -> None:
+        """Set the WorkflowRun status to COMPLETE"""
+        self._update_status(Task.COMPLETE)
+
+    def error(self) -> None:
+        """Set the WorkflowRun status to ERROR"""
+        self._update_status(Task.ERROR)
+
+    def in_progress(self) -> None:
+        """Set the WorkflowRun status to IN_PROGRESS"""
+        self._update_status(Task.IN_PROGRESS)
+
+    def execute_next_step(self) -> Optional[Task]:
```

Comment:
Should we prevent this from running if the current step hasn't finished yet? I tried running this by hand and my setup was busted so the first Task errored out. When I tried to run this on the Django shell, the json.loads in `WorkflowStep::_get_parameters()` failed because it was referencing the previous steps output.

Reply:
This is still an issue but I think it's ok since you'd have to programatically trigger the workflow to try execute the next step.